### PR TITLE
K8s-specific endpoint in agent.

### PR DESCRIPTION
**Pull request by debedb**
* Add flag to K8S specific endpoint to enable either full isolation or not
* If full isolation flag is set, specify a single policy rule in the ts-specific chain: DROP
* If full isolation flag is NOT set, specify a single policy rule in ts-specific chain: ALLOW

Also added some opportunistic changes to support things like 204 ACCEPTED (right now requests to agent just get OK in response but at that point no work is yet done), but did not finish.

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -180,7 +190,8 @@ func (a *Agent) k8sPodUpHandle(netif NetIf) error {
 		return agentError(err)
 	}
 	log.Print("Agent: provisioning firewall")
-	if err := provisionFirewallRules(netif, a); err != nil {
+
+	if err := provisionK8SFirewallRules(netReq, a, fullIsolation); err != nil {
```
Comment (flashvoid) [inline agent/agent.go]: I think `fullIsolation` is too explicit. Reader might assume there is a partial isolation and other flavours of isolation.
I'd call it just `isolation` or `enableNamespaceIsolation`.

File: agent/netconfig.go [hunk 2]
```diff
@@ -22,12 +22,24 @@ import (
...
 )
 
+const (
+	fullIsolationOption = "full_isolation"
+)
+
+// NetworkRequest specifies messages sent to the
+// agent containing information on how to configure network
+// on its host.
+type NetworkRequest struct {
```
Comment (flashvoid) [inline agent/netconfig.go]: I don't know how i feel about `NetworkRequest`, is it necessarily coming from network ?

Comment (debedb) [inline agent/netconfig.go]: I don't either, it sounded better than anything "Config" but I am not 100% happy either. Happy to name it as whatever is suggested.

File: agent/agent.go (hunk 1 above)
Comment (debedb) [inline agent/agent.go]: namespaceIsolation then. enable or not (on/off) is the value.

File: agent/agent.go [hunk 3]
```diff
@@ -159,8 +163,14 @@ func (a *Agent) index(input interface{}, ctx common.RestContext) (interface{}, e
 // 2. Ensures interhost routes are in place
 // 3. Creates ip route pointing new interface
 // 4. Provisions firewall rules
-func (a *Agent) k8sPodUpHandle(netif NetIf) error {
-	log.Print("Agent: processing request to provision new interface")
+func (a *Agent) k8sPodUpHandle(netReq NetworkRequest) error {
```
Comment (jbrendel) [inline agent/agent.go]: k8sPodUpHandle() and k8sPodUpHandler(). Very similar names.

Could we rename k8sPodUpHandle() to k8sPodUpProvision() or k8sPodUpProcessing() ?

File: agent/agent/main.go [hunk 4]
```diff
@@ -38,7 +38,7 @@ func main() {
 		fmt.Println("Must specify rootUrl.")
 		return
 	}
-	svcInfo, err := agent.Run(*rootURL)
+	svcInfo, err := agent.Run(*rootURL, false)
```
Comment (jbrendel) [inline agent/agent/main.go]: It saddens me that Go doesn't seem to have named parameters. I need to go all the way to the function definition to see what that boolean here is for. This here would be so much nicer and self-documenting:

```
svcInfo, err := agent.Run(*rootURL, testMode=false)
```

There's no decent equivalent for this in Go?

Do you really have to write:

```
svcInfo, err := agent.Run(*rootURL, false)  // flag indicates: no test mode
```

???

File: agent/firewall.go [hunk 5]
```diff
@@ -262,10 +265,16 @@ func (fw *Firewall) CreateU32Rules(chain int) error {
...
+// specified target
+func (fw *Firewall) CreateDefaultRule(chain int, target string) error {
 	log.Print("Creating default drop rules for chain", chain)
 	chainName := fw.chains[chain].chainName
 	cmd := "/sbin/iptables"
-	args := []string{"-A", chainName, "-j", "DROP"}
+	args := []string{"-A", chainName, "-j", target}
 	_, err := fw.Agent.Helper.Executor.Exec(cmd, args)
 	if err != nil {
 		log.Print("Creating default drop rules failed")
```
Comment (jbrendel) [inline agent/firewall.go]: This message here should now be: "Creating default rules failed."

File: agent/firewall.go [hunk 6]
```diff
@@ -389,6 +398,48 @@ func (fw *Firewall) extractTenantID(addr uint64) uint64 {
...
+	if err != nil {
+		return err
+	}
+	for chain := range missingChains {
+
+		//		if err := fw.CreateRules(chain); err != nil {
+		//			return err
+		//		}
+
+		var target string
```
Comment (jbrendel) [inline agent/firewall.go]: It seems to me that 'target' can be initialized outside of the for loop.

File: agent/netconfig.go (hunk 2 above)
Comment (jbrendel) [inline agent/netconfig.go]: Isn't it an endpoint-setup request?

File: common/client.go [hunk 7]
```diff
@@ -172,20 +173,21 @@ func (rc *RestClient) modifyUrl(dest string, queryMod url.Values) error {
 
 // execMethod applies the specified method to the provided url (which is interpreted
 // as relative or absolute).
+// POST methods may not be idempotent, so for retry capability we will employ the following logic:
```
Comment (jbrendel) [inline common/client.go]: Are these changes that have leaked in from some other PRs?

**Comment by jbrendel** [comment]
Done for now.

File: common/client.go (hunk 7 above)
Comment (debedb) [inline common/client.go]: Basically as agent now on a request returns "OK and then starts to do work, it reminded me it's a good idea to actually return 204, so I added some enhancements to make that possible and in the process I thought that this comment is better done as godoc.

File: agent/netconfig.go (hunk 2 above)
Comment (debedb) [inline agent/netconfig.go]: I think it's intended to be more generic than k8s endpoint... or you mean any endpoint?

File: agent/agent.go (hunk 3 above)
Comment (debedb) [inline agent/agent.go]: How about... Setup? Then we can do same for interfaceHandle (cll it interfaceSetup)

File: agent/agent/main.go (hunk 4 above)
Comment (debedb) [inline agent/agent/main.go]: no overloading funcs/methods either.

File: agent/firewall.go (hunk 5 above)
Comment (debedb) [inline agent/firewall.go]: fixed

File: agent/firewall.go (hunk 6 above)
Comment (debedb) [inline agent/firewall.go]: declared really. but fixed.

**Pull request merged**